### PR TITLE
Use Oj optionally

### PR DESCRIPTION
**Pull request by malept**
This reverts the default JSON parser to `ActiveSupport::JSON`, and only uses Oj if:

1. `ElasticRecord.json_parser` is set to `:oj`
2. Oj is loaded

This should ease upgrading.

### CODE REVIEW

File: lib/elastic_record/connection.rb [hunk 1]
```diff
@@ -38,15 +37,31 @@ def json_delete(path, json = nil)
...
-      json = Oj.compat_load(response.body)
+      json = json_decode(response.body)
       raise ConnectionError.new(response.code, json['error']) if json['error']
 
       json
     end
 
+    def json_encode(data)
+      if ElasticRecord.json_parser == :oj
+        Oj.dump(data, :mode => :compat)
```
Comment (matthuhiggins) [inline lib/elastic_record/connection.rb]: ruby 1.9 hash syntax

Comment (malept) [inline lib/elastic_record/connection.rb]: Fixed in a47c87f.

File: lib/elastic_record/index/deferred.rb [hunk 2]
```diff
@@ -42,8 +42,10 @@ def flush!
 
         private
           READ_METHODS = [:json_get, :head]
+          SERIALIZE_METHODS = [:json_encode, :json_decode]
           def method_missing(method, *args, &block)
             super unless index.real_connection.respond_to?(method)
+            return index.real_connection.send(method, *args, &block) if SERIALIZE_METHODS.include?(method)
```
Comment (matthuhiggins) [inline lib/elastic_record/index/deferred.rb]: I don't think that json_encode/json_decode should be defined on connection.rb.

File: lib/elastic_record/index/documents.rb [hunk 3]
```diff
@@ -64,7 +64,7 @@ def bulk(options = {})
         yield
 
         if current_bulk_batch.any?
-          body = current_bulk_batch.map { |action| "#{Oj.dump(action, :mode => :compat)}\n" }.join
+          body = current_bulk_batch.map { |action| "#{connection.json_encode(action)}\n" }.join
```
Comment (matthuhiggins) [inline lib/elastic_record/index/documents.rb]: Thoughts on `ElasticRecord::JSON.json_encode`?

File: lib/elastic_record.rb [hunk 4]
```diff
@@ -17,6 +17,14 @@ class << self
...
       yield(ElasticRecord::Config)
     end
+
+    def json_parser
+      @@json_parser ||= :active_support
+    end
+
+    def json_parser=(value)
+      @@json_parser = value
+    end
```
Comment (matthuhiggins) [inline lib/elastic_record.rb]: Thoughts on Thoughts on `ElasticRecord::JSON.parser = :oj`?

**Comment by malept** [comment]
> Thoughts on `ElasticRecord::JSON.encode`?
> Thoughts on `ElasticRecord::JSON.parser = :oj`?

The refactor suggestions are fine, but at this point, I'm wondering where the line is to stop reinventing the wheel and switch to `multi_json`.

**Comment by matthuhiggins** [comment]
I wish that could be an obvious suggestion, but have you read https://github.com/rails/rails/pull/10576?

**Comment by malept** [comment]
I had not, until now. I was not aware of how much overhead it is, though I shouldn't be surprised.

**Comment by malept** [comment]
Refactored in ce9370f.

**Pull request merged**